### ISSUE TITLE
Observing weird downwards jump in loss after checkpoint reloading in a DDP setting

### ISSUE THREAD

**Issue by jubueche** [huggingface/transformers#30502]
### System Info

- `transformers` version: 4.39.1
- Platform: Linux-3.10.0-1160.76.1.el7.x86_64-x86_64-with-glibc2.17
- Python version: 3.10.14
- Huggingface_hub version: 0.21.4
- Safetensors version: 0.4.2
- Accelerate version: 0.30.0.dev0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.4.0.dev20240322 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: True, 8 GPUs
- Using distributed or parallel set-up in script?: DDP

### Who can help?

@muellerz @pacman100  

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

We are observing a jump downwards in the training loss when we train a model (Roberta). This problem is best described by the two following plots:

<img width="1356" alt="Screenshot 2024-04-26 at 15 38 18" src="https://github.com/huggingface/transformers/assets/30778073/e0c3c387-9d3b-4550-a3d2-f22fbe144fd2">

We observe that the magnitude of the drop after re-loading the checkpoint is directly proportional to the amount of time of the previous run. 
Our first (and still only) guess was that there is some averaging going on. But debugging through the code on 1 GPU, I don't see any averaging effect: we add `1 / gradient_acc_steps * loss` until we call `optimizer.step()`. Afterwards, we reset the loss that we report. I don't see the issue there.

### Expected behavior

The loss should go down with smooth transitions between loading checkpoints.

**Comment by waterluck** [comment]
@jubueche Hi, jubueche , have you got some new ideas on why this happen? I meet the same problem, with loss like the pic when full fine-tune llama2:
<img width="523" alt="image" src="https://github.com/huggingface/transformers/assets/111731547/4a49b098-a0f5-4136-8388-146da2778fc3">
<img width="1088" alt="image" src="https://github.com/huggingface/transformers/assets/111731547/b927c9ad-acdf-4638-9889-07bc4b542363">

**Comment by jubueche** [comment]
No, still no idea. We had to move on and are not actively working on identifying this problem anymore :(

**Comment by waterluck** [comment]
@jubueche Thanks for the reply!  I'll keep trying to figure out.
